### ISSUE TITLE
BUG: Checking if dtype is in a set of dtypes fails, but same check with list passes.

### ISSUE THREAD

**Issue by igordertigor** [numpy/numpy#21065]
### Describe the issue:

When comparing dtypes to sets and lists of dtypes, I observe inconsistent behaviour. When checking if a dtype is in a set of dtypes, the result is always `False`. Checking against a list of dtypes this is not the case.

This might be related to #6295

### Reproduce the code example:

```python
import numpy as np

uint_types = {np.uint8, np.uint16, np.uint32}
a = np.array([0], np.uint8)

assert a.dtype in uint_types, "This fails"

# However:
assert a.dtype in list(uint_types), "This passes"
```

### NumPy/Python version information:

numpy version: 1.21.5
python version: 3.8.10 (default, Nov 26 2021, 20:14:08) \n[GCC 9.3.0]

**Comment by kidrahahjo** [comment]
The dtypes aren't intended to be singleton. I'd recommend reading https://numpy.org/neps/nep-0040-legacy-datatype-impl.html#nep40. This might help in understanding the datatype implementation in NumPy.

**Comment by matthew-brett** [comment]
I also found it confusing that set membership and list membership work in different ways:

```
[ins] In [9]: np.dtype(np.uint8) in {np.uint8}
Out[9]: False

[ins] In [10]: np.dtype(np.uint8) in [np.uint8]
Out[10]: True

[ins] In [11]: np.dtype(np.uint8) == np.uint8
Out[11]: True
```

Is it easy to explain why?

**Comment by kidrahahjo** [comment]
> I also found it confusing that set membership and list membership work in different ways:
> 
> ```
> [ins] In [9]: np.dtype(np.uint8) in {np.uint8}
> Out[9]: False
> 
> [ins] In [10]: np.dtype(np.uint8) in [np.uint8]
> Out[10]: True
> 
> [ins] In [11]: np.dtype(np.uint8) == np.uint8
> Out[11]: True
> ```
> 
> Is it easy to explain why?

For look-up on a Python `set` you need to compute hash values but with a `list` you will end up doing a linear search.
It may be possible that the `__hash__` function of class `np.dtype` is computed in a different way than the class `np.uint8`. Hence, the different outputs for your queries.

**Comment by matthew-brett** [comment]
Aha - yes:

```
[ins] In [5]: hash(np.dtype(np.uint8)) == hash(np.uint8)                                                                           
Out[5]: False
```

But - that seems wrong to me.  Quoting from the [Python documentation for hash](https://docs.python.org/3/library/functions.html#hash):

> Numeric values that compare equal have the same hash value (even if they are of different types, as is the case for 1 and 1.0)

I believe that the contract for `hash` is that the returned values should always be the same for two values that compare equal, but there may be objects that do not compare equal that return the same hash.

**Comment by kidrahahjo** [comment]
You're right! This seems like a bug to me as well. Great find. I can work to fix this issue if it's really a bug and not something intentional.

**Comment by igordertigor** [comment]
I would also say, it's a bug, for similar reasons as @matthew-brett. I understand the reasoning behind NEP 40, but my understanding is also that this is not something that should stay for all eternity. Practically, I can of course work with dtypes not being hashable.

**Comment by kidrahahjo** [comment]
Can we implement a top-level MetaClass for the class `dtype` for this purpose?

**Comment by seberg** [comment]
> Can we implement a top-level MetaClass for the class dtype for this purpose?

We already have, I was talking about the *scalar* types.  I have to admit, I am not sure if this is easy fixed to begin with the fact that `np.dtype(...) = ...` evaluates to `np.dtype(...) == np.dtype(...)` is to some degree just broken, probably.

But yeah, there should be ways to make it more consistent.
